File: java/src/main/java/com/fidelity/promptlab/services/UserService.java
package com.fidelity.promptlab.services;

import com.fidelity.promptlab.models.*;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * This service has intentional issues for participants to improve with good prompts.
 */
public class UserService {

    private final Map<String, User> users = new HashMap<>();

    // Challenge: This login method is incomplete and insecure
    // Participants will be asked to improve it with prompts
    public Object login(String email, String password) {
        // TODO: Implement proper authentication
        // Current implementation is intentionally weak
        if (email != null && password != null) {
            Map<String, Object> result = new HashMap<>();
            result.put("success", true);
            result.put("token", "fake-token");
            return result;
        }
        Map<String, Object> result = new HashMap<>();
        result.put("success", false);
        return result;
    }

    // Challenge: No input validation
    public User createUser(Map<String, Object> data) {
        User user = new User();
        user.setId(UUID.randomUUID().toString());
        user.setEmail((String) data.get("email"));
        user.setFirstName((String) data.get("firstName"));
        user.setLastName((String) data.get("lastName"));
        user.setRole(UserRole.CUSTOMER);
        user.setPreferences(new UserPreferences());

        users.put(user.getId(), user);
        return user;
    }

    // Challenge: No error handling - returns null if not found
    public User getUser(String id) {
        return users.get(id);
    }

    // Challenge: Needs optimization - currently O(n)
    public User findByEmail(String email) {
        for (User user : users.values()) {
            if (user.getEmail().equals(email)) {
                return user;
            }
        }
        return null;
    }

    // Challenge: Missing audit logging
    public User updateUser(String id, Map<String, Object> updates) {
        User user = users.get(id);
        if (user == null) {
            throw new RuntimeException("User not found");
        }

        if (updates.containsKey("email")) {
            user.setEmail((String) updates.get("email"));
        }
        if (updates.containsKey("firstName")) {
            user.setFirstName((String) updates.get("firstName"));
        }
        if (updates.containsKey("lastName")) {
            user.setLastName((String) updates.get("lastName"));
        }

        return user;
    }

    // Challenge: Hard delete with no soft delete option
    public void deleteUser(String id) {
        users.remove(id);
    }
}

File: angular/src/services/user.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Observable, of, throwError } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { User, LoginRequest, LoginResponse, Result, ApiError } from '../models/types';

/**
 * This service has intentional issues for participants to improve with good prompts.
 *
 * INTENTIONAL ISSUES:
 * 1. Login returns 'any' instead of proper LoginResponse
 * 2. No input validation on createUser
 * 3. No error handling on getUser (silent failures)
 * 4. O(n) email lookup instead of using a Map or API endpoint
 * 5. Missing audit logging on updateUser
 * 6. Hard delete with no soft delete option
 * 7. No retry logic on API calls
 * 8. No caching strategy
 */
@Injectable({ providedIn: 'root' })
export class UserService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = '/api/users';

  // Challenge: This login method is incomplete and insecure
  // - Returns 'any' instead of Observable<LoginResponse>
  // - No input validation
  // - No rate limiting consideration
  // - No secure token storage strategy
  login(email: string, password: string): Observable<any> {
    // TODO: Implement proper authentication
    // Current implementation is intentionally weak
    if (email && password) {
      return of({ success: true, token: 'fake-token' });
    }
    return of({ success: false });
  }

  // Challenge: No input validation, no proper typing
  // - Accepts 'any' instead of proper DTO
  // - No email format validation
  // - No password strength check
  // - No duplicate email check
  createUser(data: any): Observable<User> {
    return this.http.post<User>(this.apiUrl, data);
  }

  // Challenge: No error handling
  // - Returns undefined if not found (should throw or return Observable<User | null>)
  // - No loading state
  // - No retry on failure
  getUser(id: string): Observable<User> {
    return this.http.get<User>(`${this.apiUrl}/${id}`);
    // Bug: No error handling - will throw to subscriber
  }

  // Challenge: Needs optimization
  // - Makes API call to get ALL users then filters (should use query param)
  // - No caching
  // - O(n) client-side filtering
  findByEmail(email: string): Observable<User | undefined> {
    return this.http.get<User[]>(this.apiUrl).pipe(
      map(users => users.find(user => user.email === email))
      // Bug: Fetches all users just to find one by email
    );
  }

  // Challenge: Missing audit logging
  // - No tracking of what changed
  // - No user who made the change
  // - No timestamp of change
  updateUser(id: string, updates: Partial<User>): Observable<User> {
    return this.http.patch<User>(`${this.apiUrl}/${id}`, updates);
  }

  // Challenge: Hard delete with no soft delete option
  // - No confirmation
  // - No undo capability
  // - No cascade handling
  deleteUser(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
